openmc/weight_windows.py

```diff
@@ -410,3 +406,170 @@ def from_hdf5(cls, group, meshes):
+        else:
+            raise RuntimeError(f'Invalid mesh description (nr) found: {nr}')
+
+        # read BLOCK 2 and BLOCK 3 data into a single array
+        ww_data = np.fromstring(wwinp.read(), sep=' ')
+
+    # extract mesh data from the ww_data array
+    start_idx = 0
+
+    end_idx = start_idx + 1 + 3 * ncx
+    x_vals = ww_data[start_idx:end_idx]
+    start_idx = end_idx
+
+    end_idx = start_idx + 1 + 3 * ncy
+    y_vals = ww_data[start_idx:end_idx]
+    start_idx = end_idx
+
+    end_idx = start_idx + 1 + 3 * ncz
+    z_vals = ww_data[start_idx:end_idx]
+    start_idx = end_idx
+
+    # mesh consistency checks
+    if nr == 16 and nwg == 1:
+        raise ValueError(f'Mesh description in header ({nr}) '
+                         f'does not match the mesh type ({nwg})')
+
+    if not (xyz0 == (x_vals[0], y_vals[0], z_vals[0])).all():
+        raise ValueError(f'Mesh origin in the header ({xyz0}) '
+                         f' does not match the origin in the mesh '
+                         f' description ({x_vals[0], y_vals[0], z_vals[0]})')
+
+    # extract weight window values from array
+    particle_types = {0: 'neutron', 1: 'photon'}
+    particle_data = []
+    for p in range(ni):
+        # no information to read for this particle if
+        # either the energy bins or time bins are empty
+        if ne[p] == 0 or nt[p] == 0:
+            continue
+
+        if iv > 1:
+            # time bins are parsed but unused for now
+            end_idx = start_idx + nt[p]
+            time_bounds = ww_data[start_idx:end_idx]
+            np.insert(time_bounds, (0,), (0.0,))
+            start_idx = end_idx
+
+        # read energy boundaries
+        end_idx = start_idx + ne[p]
+        energy_bounds = np.insert(ww_data[start_idx:end_idx], (0,), (0.0,))
+        # convert from MeV to eV
+        energy_bounds *= 1e6
+        start_idx = end_idx
+
+        # read weight window values
+        end_idx = start_idx + (nfx * nfy * nfz) * nt[p] * ne[p]
+
+        # read values and reshape according to ordering
+        # slowest to fastest: z, y, x, e
+        ww_values = ww_data[start_idx:end_idx].reshape(nfz, nfy, nfx, ne[p])
+        # swap z and x axes for correct shape and (ijk, e) mesh indexing
+        ww_values = np.swapaxes(ww_values, 2, 0)
+        start_idx = end_idx
+
+        # gather particle data in container
+        particle_dict = {'particle': particle_types[p],
+                         'energy_bounds': energy_bounds,
+                         'ww_values': ww_values}
+        particle_data.append(particle_dict)
+
+    # create openmc mesh object
+    grids = []
+    for grid_info in [x_vals, y_vals, z_vals]:
+        # first value is the start of the grid
+        coords = [grid_info[0]]
+
+        # extend the grid based on the next coarse bin endpoint, px
+        # and the number of fine bins in the coarse bin, sx
+        intervals = grid_info[1:].reshape(-1, 3)
```

Comment:
small nitpick, take it or leave it, - you could move this reshaping up to the previous section in order to do the checks on "s" and "q" more easily something like `x0, xvals = xvals[0], xvals[1:].reshape(-1, 3)`

Reply:
I like it. More concise. 👍🏻